<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="LucianoRogerio" />

<meta name="date" content="2022-01-31" />

<title>Lab2 Luciano Solution</title>

<script src="site_libs/header-attrs-2.11/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">QuantitativeGen</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Luciano Solutions
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="LucSolutions.html">Home</a>
    </li>
    <li>
      <a href="Lab1.html">Homework 1</a>
    </li>
    <li>
      <a href="Lab2.html">Homework 2</a>
    </li>
  </ul>
</li>
<li>
  <a href="LabClasses.html">LabClasses</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Lab2 Luciano Solution</h1>
<h4 class="author">LucianoRogerio</h4>
<h4 class="date">2022-01-31</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-02-01
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>QuantitativeGen/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version 1.7.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20220124code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20220124)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20220124code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20220124)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomLucianoRogerioQuantitativeGenLabLuctree79038f79048548f2c55242d7e4538ee0d4be72ebtargetblank79038f7a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/LucianoRogerio/QuantitativeGenLabLuc/tree/79038f79048548f2c55242d7e4538ee0d4be72eb" target="_blank">79038f7</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomLucianoRogerioQuantitativeGenLabLuctree79038f79048548f2c55242d7e4538ee0d4be72ebtargetblank79038f7a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/LucianoRogerio/QuantitativeGenLabLuc/tree/79038f79048548f2c55242d7e4538ee0d4be72eb" target="_blank">79038f7</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  output/QTLgenotypes.csv

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/Lab2.Rmd</code>) and HTML (<code>docs/Lab2.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/LucianoRogerio/QuantitativeGenLabLuc/blob/79038f79048548f2c55242d7e4538ee0d4be72eb/analysis/Lab2.Rmd" target="_blank">79038f7</a>
</td>
<td>
LucianoRogerio
</td>
<td>
2022-02-01
</td>
<td>
Homework 2
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/LucianoRogerio/QuantitativeGenLabLuc/79038f79048548f2c55242d7e4538ee0d4be72eb/docs/Lab2.html" target="_blank">79038f7</a>
</td>
<td>
LucianoRogerio
</td>
<td>
2022-02-01
</td>
<td>
Homework 2
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="menu" />


<div id="homework---jan-31-2022" class="section level1">
<h1>Homework - Jan 31 2022</h1>
<p>Illustrate regression to the mean between parents and offspring using AlphaSimR<br />
1. You know how to create a population – create a population of parents. <a href="#P1">Answer</a><br />
2. You know how to get phenotypes from that population. <a href="#P2">Answer</a><br />
3. You want to randomly mate that population to get progeny that will be regressed to the mean. Create a <code>crossPlan</code> matrix like I did to generate the F1s, except that each row should have randomly-picked parents, rather than 1 and 2 like for the F1s. There is also a command <code>AlphaSimR::randCross</code>. Check out its documentation. If you use that command, you will have to find the pedigree of the progeny using the <code>@mother</code> and <code>@father</code> ids of the progeny population and match those up to the parent population. If you make the crossPlan, then it gives you the seed and pollen parent ids. <a href="#P3">Answer</a><br />
4. Having made the progeny population, phenotype it also. <a href="#P4">Answer</a><br />
5. Use each row of the <code>crossPlan</code> to find the two parents and calculate their phenotypic mean. <a href="#P5">Answer</a><br />
6. Make a scatterplot of the progeny phenotypes against the parent mean phenotypes. <a href="#P6">Answer</a></p>
<blockquote>
<p>What is the regression coefficient?</p>
</blockquote>
<ul>
<li>Regression coefficient in this case is the broad sense heritability index of the simulated trait</li>
</ul>
<blockquote>
<p>How much closer to the mean are offspring, on average, compared to their parents?</p>
</blockquote>
<ul>
<li>The offspring mean observed were 0.42 of the parent mean in one of my Simulations, but this value keeps changing if I run all again, even using the set.seed.</li>
<li>I was expecting 0.5 that was the simulation parameter added to broad sense heritability, <a href="#P7">results</a></li>
</ul>
<div id="checking-if-the-packages-required-are-installed" class="section level3">
<h3>Checking if the packages required are installed</h3>
<pre class="r"><code>packages_used &lt;- c(&quot;AlphaSimR&quot;, &quot;tidyverse&quot;, &quot;workflowr&quot;, &quot;here&quot;)
ip &lt;- installed.packages()
all_packages_installed &lt;- TRUE
for (package in packages_used){
  if (!(package %in% ip[, &quot;Package&quot;])){
    print(paste(&quot;Please install package&quot;, package, sep = &quot; &quot;))
    all_packages_installed &lt;- FALSE
  }
}#END packages_used
if (!all_packages_installed) stop(&quot;Need to install packages&quot;)

library(tidyverse); library(here)</code></pre>
<pre><code>── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──</code></pre>
<pre><code>✓ ggplot2 3.3.5     ✓ purrr   0.3.4
✓ tibble  3.1.6     ✓ dplyr   1.0.7
✓ tidyr   1.1.4     ✓ stringr 1.4.0
✓ readr   2.1.1     ✓ forcats 0.5.1</code></pre>
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
x dplyr::filter() masks stats::filter()
x dplyr::lag()    masks stats::lag()</code></pre>
<pre><code>here() starts at /Users/lbd54/Documents/GitHub/QuantitativeGen</code></pre>
<p>loading the here package it already fix the Project directory as the working directory.</p>
</div>
<div id="loading-the-packages-and-creating-a-readme.rmd-to-add-information-of-software-used-in-this-analysis" class="section level3">
<h3>Loading the packages and creating a README.Rmd to add information of software used in this analysis</h3>
<pre class="r"><code>source(here::here(&quot;code/addToREADME.R&quot;))
addToREADME(paste0(&quot;## &quot;, rmarkdown::metadata$title), append = F)
addToREADME(c(date(), &quot;&quot;))

packages_info &lt;- ip[packages_used, c(&quot;Package&quot;, &quot;Version&quot;, &quot;Built&quot;)]
addToREADME(c(&quot;The packages used in this script are:&quot;, &quot;Package, Version, Built&quot;))
apply(packages_info, 1, function(vec) addToREADME(paste(vec, collapse = &quot; &quot;)))</code></pre>
<pre><code>              AlphaSimR               tidyverse               workflowr 
&quot;AlphaSimR 1.0.4 4.1.1&quot; &quot;tidyverse 1.3.1 4.1.0&quot; &quot;workflowr 1.7.0 4.1.1&quot; 
                   here 
     &quot;here 1.0.1 4.1.0&quot; </code></pre>
<div id="P1" />

</div>
<div id="seed-set-and-simulations-parameters" class="section level3">
<h3>Seed set and Simulations parameters</h3>
<pre class="r"><code>set.seed(1312022)

addToREADME(&quot;&quot;)

addToREADME(paste(&quot;set.seed =&quot;, &quot;1312022&quot;, sep = &quot; &quot;))
addToREADME(&quot;&quot;)

# Simulations parameters for the population
nFounders &lt;- 100
inbred &lt;- FALSE
ploidy &lt;- 2L
nChr &lt;- 18
effPopSize &lt;- 200
segSites &lt;- 1650
nQTL &lt;- 1000
nSNP &lt;- 500
nCores &lt;- 4


addToREADME(c(paste(&quot;Number of Founders --&gt;&quot;, nFounders, sep = &quot; &quot;),
              paste(&quot;Is the founders inbred?&quot;, inbred, sep = &quot; &quot;),
              paste(&quot;Which is the ploidy?&quot;, ploidy, sep = &quot; &quot;),
              paste(&quot;Number of Chromosomes --&gt;&quot;, nChr, sep = &quot; &quot;),
              paste(&quot;Size of the effective population --&gt;&quot;, effPopSize, sep = &quot; &quot;),
              paste(&quot;Number of quatitative loci traits per chromosome expected affecting the trait --&gt;&quot;, nQTL, sep = &quot; &quot;),
              paste(&quot;Number of SNP markers per chromosome --&gt;&quot;, nSNP, sep = &quot; &quot;),
              paste(&quot;Number of cores to apply parallel to the simulations of the chromosomes --&gt;&quot;, nCores, sep = &quot; &quot;)))

addToREADME(&quot; &quot;)</code></pre>
<div id="P2" />


</div>
<div id="creating-the-founder-population-and-the-parents-population" class="section level3">
<h3>Creating the Founder population and the parents population</h3>
<pre class="r"><code>founderPop &lt;- AlphaSimR::runMacs(nInd = nFounders,
                                 inbred = inbred,
                                 ploidy = ploidy,
                                 nChr = nChr,
                                 segSites = segSites,
                                 nThreads = nCores)

addToREADME(&quot;Founder Pop Info&quot;)
addToREADME(&quot;----------------&quot;)
capture.output(founderPop, file = here::here(&quot;docs/README.md&quot;), type = &quot;output&quot;, append = TRUE)
addToREADME(&quot; &quot;)

SP &lt;- AlphaSimR::SimParam$new(founderPop = founderPop)$addTraitA(nQtlPerChr = nQTL)$setVarE(H2 = 0.5)

# Simulating the Founder Pop with phenotype
foundersID &lt;- AlphaSimR::newPop(rawPop = founderPop,
                                    simParam = SP)

addToREADME(&quot;Founders Ids Info&quot;)
addToREADME(&quot;----------------&quot;)
capture.output(foundersID, file = here::here(&quot;docs/README.md&quot;), type = &quot;output&quot;, append = TRUE)
addToREADME(&quot; &quot;)</code></pre>
<div id="P3" />

</div>
<div id="simulating-the-f1-and-f2-populations" class="section level3">
<h3>Simulating the F1 and F2 populations</h3>
<pre class="r"><code>F1 &lt;- AlphaSimR::selectCross(pop = foundersID,
                             nInd = 35,
                             nCrosses = 200,
                             use = &quot;pheno&quot;)
F1 &lt;- AlphaSimR::setPheno(F1)
ParentsF1ID &lt;- c(F1@mother,F1@father) %&gt;% unique %&gt;% .[order(as.numeric(.))]
ParentsF1 &lt;- AlphaSimR::selectInd(foundersID, nInd = 35, candidates = ParentsF1ID)
ParentsF1 &lt;- AlphaSimR::setPheno(ParentsF1)
F2 &lt;- AlphaSimR::self(F1)
F2 &lt;- AlphaSimR::setPheno(F2)

addToREADME(&quot;Populations Sim&quot;)
addToREADME(&quot;---------------&quot;)
addToREADME(&quot;Parents F1 Selected&quot;)
addToREADME(&quot;-------------------&quot;)
capture.output(ParentsF1, file = here::here(&quot;docs/README.md&quot;), type = &quot;output&quot;, append = TRUE)
addToREADME(&quot; &quot;)
addToREADME(&quot;F1 Simulated&quot;)
addToREADME(&quot;------------&quot;)
capture.output(F1, file = here::here(&quot;docs/README.md&quot;), type = &quot;output&quot;, append = TRUE)
addToREADME(&quot; &quot;)
addToREADME(&quot;F2 Simulated&quot;)
addToREADME(&quot;------------&quot;)
capture.output(F2, file = here::here(&quot;docs/README.md&quot;), type = &quot;output&quot;, append = TRUE)
addToREADME(&quot; &quot;)</code></pre>
<div id="P4" />

</div>
<div id="obtain-the-data-from-each-population" class="section level3">
<h3>Obtain the data from each Population</h3>
<pre class="r"><code>ParentsF1Dt &lt;- data.frame(ID = ParentsF1@id,
                          pheno = ParentsF1@pheno)

F1Dt &lt;- data.frame(ID = F1@id,
                   mother = F1@mother,
                   father = F1@father,
                   pheno = F1@pheno)

F2Dt &lt;- data.frame(ID = F2@id,
                   mother = F2@mother,
                   father = F2@father,
                   pheno = F2@pheno)</code></pre>
<div id="P5" />

</div>
<div id="prepare-the-pop-data-with-the-pheno-of-the-parents-in-the-same-file" class="section level3">
<h3>Prepare the Pop data with the Pheno of the parents in the same file</h3>
<pre class="r"><code>F1Dt2 &lt;- left_join(F1Dt, ParentsF1Dt, by = c(&quot;mother&quot; = &quot;ID&quot;), suffix = c(&quot;&quot;, &quot;.mother&quot;)) %&gt;%
  left_join(., ParentsF1Dt, by = c(&quot;father&quot; = &quot;ID&quot;), suffix = c(&quot;&quot;, &quot;.father&quot;)) %&gt;%
  dplyr::mutate(pheno.ParMean = (pheno.mother + pheno.father)/2,
                Pop = &quot;F1&quot;)

F2Dt2 &lt;- left_join(F2Dt, ParentsF1Dt, by = c(&quot;mother&quot; = &quot;ID&quot;), suffix = c(&quot;&quot;, &quot;.mother&quot;)) %&gt;%
  left_join(., ParentsF1Dt, by = c(&quot;father&quot; = &quot;ID&quot;), suffix = c(&quot;&quot;, &quot;.father&quot;)) %&gt;%
  dplyr::mutate(pheno.ParMean = (pheno.mother + pheno.father)/2,
                Pop = &quot;F2&quot;)

Data &lt;- rbind(F1Dt2, F2Dt2)</code></pre>
<div id="P6" />

</div>
<div id="figure-of-the-regression-of-the-f1-and-f2-progeny-by-it-parent-mean" class="section level3">
<h3>Figure of the regression of the F1 and F2 progeny by it parent mean</h3>
<pre class="r"><code>Data %&gt;% 
  ggplot(aes(x = pheno.ParMean, y = pheno, color = Pop)) +
  geom_point() +
  geom_smooth(method = &quot;lm&quot;, alpha = 0.25) +
  geom_abline(intercept = 0, slope = 1, color = &quot;red&quot;) +
  theme_bw() + xlab(&quot;Parent Mean&quot;) + ylab(&quot;Progeny&quot;)</code></pre>
<pre><code>`geom_smooth()` using formula &#39;y ~ x&#39;</code></pre>
<p><img src="figure/Lab2.Rmd/Figure-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-Figure-1">
Past versions of Figure-1.png
</button>
</p>
<div id="fig-Figure-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/LucianoRogerio/QuantitativeGenLabLuc/blob/79038f79048548f2c55242d7e4538ee0d4be72eb/docs/figure/Lab2.Rmd/Figure-1.png" target="_blank">79038f7</a>
</td>
<td>
LucianoRogerio
</td>
<td>
2022-02-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="P7" />

</div>
<div id="estimation-of-the-regression-coefficient" class="section level3">
<h3>Estimation of the regression coefficient</h3>
<pre class="r"><code>Data %&gt;% filter(Pop == &quot;F1&quot;) %&gt;%
  lm(formula = .$pheno ~ .$&#39;pheno.ParMean&#39;) %&gt;%
  coefficients %&gt;% .[1] %&gt;% round(., digits = 2) %&gt;%
  paste(&quot;The observed broad sense heritability is&quot;, .,
        &quot;while the expected was&quot;, SP$varG/(SP$varG + SP$varE), sep = &quot; &quot;)</code></pre>
<pre><code>[1] &quot;The observed broad sense heritability is 0.49 while the expected was 0.5&quot;</code></pre>
<p><a href="#menu">Menu</a></p>
<p><a href="LucSolutions.html">Home</a></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.1.1 (2021-08-10)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Big Sur 11.6.1

Matrix products: default
LAPACK: /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] here_1.0.1      forcats_0.5.1   stringr_1.4.0   dplyr_1.0.7    
 [5] purrr_0.3.4     readr_2.1.1     tidyr_1.1.4     tibble_3.1.6   
 [9] ggplot2_3.3.5   tidyverse_1.3.1

loaded via a namespace (and not attached):
 [1] httr_1.4.2       sass_0.4.0       bit64_4.0.5      vroom_1.5.7     
 [5] jsonlite_1.7.3   splines_4.1.1    modelr_0.1.8     bslib_0.3.1     
 [9] assertthat_0.2.1 highr_0.9        cellranger_1.1.0 yaml_2.2.1      
[13] pillar_1.6.4     backports_1.4.1  lattice_0.20-45  glue_1.6.0      
[17] digest_0.6.29    promises_1.2.0.1 rvest_1.0.2      colorspace_2.0-2
[21] htmltools_0.5.2  httpuv_1.6.5     Matrix_1.4-0     pkgconfig_2.0.3 
[25] broom_0.7.11     haven_2.4.3      scales_1.1.1     whisker_0.4     
[29] later_1.3.0      tzdb_0.2.0       git2r_0.29.0     mgcv_1.8-38     
[33] farver_2.1.0     generics_0.1.1   ellipsis_0.3.2   withr_2.4.3     
[37] AlphaSimR_1.0.4  cli_3.1.0        magrittr_2.0.1   crayon_1.4.2    
[41] readxl_1.3.1     evaluate_0.14    fs_1.5.2         fansi_1.0.2     
[45] nlme_3.1-153     xml2_1.3.3       tools_4.1.1      hms_1.1.1       
[49] lifecycle_1.0.1  munsell_0.5.0    reprex_2.0.1     compiler_4.1.1  
[53] jquerylib_0.1.4  rlang_0.4.12     grid_4.1.1       rstudioapi_0.13 
[57] labeling_0.4.2   rmarkdown_2.11   gtable_0.3.0     DBI_1.1.2       
[61] R6_2.5.1         lubridate_1.8.0  knitr_1.37       fastmap_1.1.0   
[65] bit_4.0.4        utf8_1.2.2       workflowr_1.7.0  rprojroot_2.0.2 
[69] stringi_1.7.6    parallel_4.1.1   Rcpp_1.0.8       vctrs_0.3.8     
[73] dbplyr_2.1.1     tidyselect_1.1.1 xfun_0.29       </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
